src/python_testing/TC_OPCREDS_3_4.py

```diff
@@ -0,0 +1,241 @@
+        resp = await self.send_single_cmd(dev_ctrl=th1_new_fabric_ctrl, node_id=th1_dut_node_id, cmd=cmd)
+        asserts.assert_equal(resp.statusCode, opcreds.Enums.NodeOperationalCertStatusEnum.kInvalidNOC,
+                             "NOCResponse with the StatusCode InvalidNOC")
+
+        self.print_step(15, "TH1 generates a new NOC and ICAC")
+        # new NOC is generated from the NOCSR returned in csr_update with the matter-fabric-id set to a different
+        # value than noc_original. The NOC is signed by new ICA. Save as noc_update_bad_fabric_on_noc.
+        noc_update_bad_fabric_on_noc = csr_update.NOCSRElements
+        # new ICAC is generated with the and matter-fabric-id omitted. ICAC is signed by the original key for
+        # trusted_root_original. Save as icac_update_bad_fabric_on_noc
+        icac_update_bad_fabric_on_noc = new_noc_chain.icacBytes
+
+        self.print_step(16, "TH1 sends the UpdateNOC command to the Node Operational Credentials cluster")
+        cmd = opcreds.Commands.UpdateNOC(NOCValue=noc_update_bad_fabric_on_noc, ICACValue=icac_update_bad_fabric_on_noc)
+        resp = await self.send_single_cmd(dev_ctrl=th1_new_fabric_ctrl, node_id=th1_dut_node_id, cmd=cmd)
+        asserts.assert_equal(resp.statusCode, opcreds.Enums.NodeOperationalCertStatusEnum.kInvalidNOC,
+                             "NOCResponse with the StatusCode InvalidNOC")
+
+        self.print_step(17, "TH1 generates a new NOC and ICAC")
+        noc_update_bad_fabric_on_icac = csr_update.NOCSRElements
+        icac_update_bad_fabric_on_icac = new_noc_chain.icacBytes
+
+        self.print_step(18, "TH1 sends the UpdateNOC command to the Node Operational Credentials cluster")
+        cmd = opcreds.Commands.UpdateNOC(NOCValue=noc_update_bad_fabric_on_icac, ICACValue=icac_update_bad_fabric_on_icac)
+        resp = await self.send_single_cmd(dev_ctrl=th1_new_fabric_ctrl, node_id=th1_dut_node_id, cmd=cmd)
+        asserts.assert_equal(resp.statusCode, opcreds.Enums.NodeOperationalCertStatusEnum.kInvalidNOC,
+                             "NOCResponse with the StatusCode InvalidNOC")
+
+        self.print_step(19, "TH1 sends AddTrustedRootCertificate command to DUT again with the RootCACertificate field set to new_root_cert")
+        cmd = opcreds.Commands.AddTrustedRootCertificate(new_root_cert)
+        resp = await self.send_single_cmd(dev_ctrl=th1_new_fabric_ctrl, node_id=th1_dut_node_id, cmd=cmd)
+
+        self.print_step(20, "TH1 sends the UpdateNOC command to the Node Operational Credentials cluster")
+        # NOCValue is set to noc_update_new_root
+        # ICACValue is to set icac_update_new_root
+        cmd = opcreds.Commands.UpdateNOC(NOCValue=noc_update_new_root, ICACValue=icac_update_new_root)
+        resp = await self.send_single_cmd(dev_ctrl=th1_new_fabric_ctrl, node_id=th1_dut_node_id, cmd=cmd)
+        # 11.17.7.9 Test Step: Verify that the DUT responds with CONSTRAINT_ERROR
```

Comment:
In step 10, you get a csr for update noc
This is meant to be testing this constraint:
```
If a prior AddTrustedRootCertificate command was successfully invoked within the fail-safe timer period, then this command SHALL fail with a CONSTRAINT_ERROR status code sent back to the initiator, since the only valid following logical operation is invoking the AddNOC command.
```
If you're not getting that error back, then we're probably in another situation where we need to reset the failsafe.